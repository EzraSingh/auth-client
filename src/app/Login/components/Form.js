import React, { Component } from 'react'
import { Redirect, Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { login } from 'actions/auth'

class LoginForm extends Component{
    state = { reveal: false }
    handleSubmit = event => {
        event.preventDefault()
        const data = new FormData(event.target)
        this.props.login({ 
            email : data.get('email'),
            password : data.get('password')
        })
    }
    toggleVisibility = () => {
        this.setState({ reveal: !this.state.reveal })
    }
    render(){
        if(this.props.isLoggedIn){
            return <Redirect to='/profile'/>
        }
        return(
            <form onSubmit={this.handleSubmit}>
                <header>
                    <h3>Login</h3>
                    <p>Authenticate identity</p>
                </header>
                <label htmlFor="email">Email</label>
                <input id="email" name="email" type="email" required/>

                <label htmlFor="password">Password</label>
                <input id="password" name="password" type={this.state.reveal ? "text" : "password"} required/>
                <div className="hide-show">
                    <span onClick={this.toggleVisibility}>
                        {this.state.reveal ? 'Hide' : 'Show'}
                    </span>
                </div>
                <input className="btn" type="submit" value="Login"/>
                <div className="prompt">
                    <Link to='/register'>Don't have an account?</Link>
                </div>
            </form>
        )
    }
}

const mapStateToProps = state => {
    return { ...state.auth }
}

const mapDispatchToProps = dispatch => {
    return { login: credentials => dispatch(login(credentials)) }
}

export { LoginForm }
export default connect(mapStateToProps, mapDispatchToProps)(LoginForm)