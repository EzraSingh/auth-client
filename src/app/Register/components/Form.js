import React, { Component } from 'react'
import { Redirect } from 'react-router-dom'
import { connect } from 'react-redux'
import { register } from 'actions/user'

class RegisterForm extends Component{
    state = { redirect: false }
    handleSubmit = (event) => {
        event.preventDefault()
        const data = new FormData(event.target)
        this.props.register({
            email : data.get('email'),
            password : data.get('password'),
            confirm : data.get('confirm')
        })
    }
    render(){
        if(this.state.redirect){
            return <Redirect exact push to='/'/>
        }
        return(
            <form onSubmit={this.handleSubmit}>
                <label htmlFor="email">Email:</label>
                <input name="email" id="email" type="email" required/>

                <label htmlFor="password">Password:</label>
                <input name="password" id="password" type="password" required/>
                
                <label htmlFor="confirm">Confirm Password:</label>
                <input name="confirm" id="confirm" type="password" required/>

                <input type="submit" value="Create Account"/>
            </form>
        )
    }
}

const mapDispatchToProps = dispatch => {
    return { register : credentials => dispatch(register(credentials)) }
}

export default connect(null, mapDispatchToProps)(RegisterForm)