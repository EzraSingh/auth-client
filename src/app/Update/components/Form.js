import React, { Component } from 'react'
import { connect } from 'react-redux'
import { isEmpty } from 'lodash'
import { update } from 'actions/user'

class UpdateProfile extends Component{
    state = { redirect: false }
    handleSubmit = event => {
        event.preventDefault()
        const data = new FormData(event.target)
        this.props.update({
            name : data.get('name')
        })
    }
    render(){
        const { profile } = this.props
        return(
            <form onSubmit={this.handleSubmit}>
                <label htmlFor="name">Name</label>
                <input id="name" name="name" placeholder={!isEmpty(profile) && "name" in profile && profile.name}/>

                <input type="submit" value="Save"/>
            </form>
        )
    }
}

const mapStateToProps = state => {
    return { ...state.user }
}

const mapDispatchToProps = dispatch => {
    return { update : profile => dispatch(update(profile)) }
}

export default connect(mapStateToProps, mapDispatchToProps)(UpdateProfile)