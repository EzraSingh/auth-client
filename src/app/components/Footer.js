import React from 'react'

const Footer = () => {
    return(
        <footer>
            By 
            <a href="https://github.com/EzraSingh" target="_BLANK" rel="noopener noreferrer">
                Ezra Singh,
            </a>
            browse the source on
            <a href="https://github.com/EzraSingh/auth-client" target="_BLANK" rel="noopener noreferrer">
                <i className="fab fa-github"/>
            </a>
        </footer>
    )
}

export default Footer