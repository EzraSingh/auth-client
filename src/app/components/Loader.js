import React from 'react'

const Loader = () => {
    return <div className='sk-rotating-plane'/>
}

export default Loader