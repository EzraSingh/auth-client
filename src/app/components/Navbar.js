import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { NavLink } from 'react-router-dom'
import { connect } from 'react-redux'

/** Ruute template for registering routes within NavBar  */
function Route(name, href, auth, hideOnAuth){
    this.name = name
    this.href = href
    this.auth = auth
    this.hideOnAuth = hideOnAuth
}

class Navbar extends Component{
    static propTypes = {
        isLoggedIn: PropTypes.bool.isRequired,
        avatar: PropTypes.string,
        name: PropTypes.string
    }
    static routes = [
        new Route('Logout', '/logout', true)
    ]
    /** Filter routes that require authentication */
    generateRoutes(){
        if(!this.props.isLoggedIn){
            // Filter routes that do not require auth
            return Navbar.routes.filter((route) => !route.auth)
        }
        else{
            // Filter routes that require auth and are not flagged to hide on auth
            return Navbar.routes.filter((route) => route.auth && !route.hideOnAuth)
        }
    }
    renderAvatar = () =>{
        return(
            <figure className="avatar animated rotateIn delay-2">
                <NavLink to="/profile">
                    <img 
                        src={this.props.avatar || process.env.REACT_APP_DEFAULT_AVATAR} 
                        alt={this.props.name}
                    />
                </NavLink>
            </figure>
        )
    }
    render(){
        return(
            <nav className="flex navbar animated slideInDown">
                {this.generateRoutes().map((route) => (
                    <NavLink className="button button-outline" to={route.href} key={route.href}>
                        {route.name}
                    </NavLink>
                ))}
                {this.props.isLoggedIn && this.renderAvatar()}
            </nav>
        )
    }
}

const mapStateToProps = state => {
    return { ...state.auth, avatar: state.user.profile.avatar, name: state.user.profile.name }
}

export { Navbar as NavbarComponent }
export default connect(mapStateToProps)(Navbar)