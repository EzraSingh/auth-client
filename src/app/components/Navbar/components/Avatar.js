import React from 'react'
import PropTypes from 'prop-types'
import { NavLink } from 'react-router-dom'

const Avatar = ({ avatar, name }) => {
    return(
        <NavLink id="avatar" to="/profile">
            <figure className="is-64 is-round">
                <img
                    src={avatar || process.env.REACT_APP_DEFAULT_AVATAR} 
                    alt={name}
                />
            </figure>
        </NavLink>
    )
}

Avatar.propTypes = {
    avatar: PropTypes.string,
    name: PropTypes.string.isRequired
}

export default Avatar