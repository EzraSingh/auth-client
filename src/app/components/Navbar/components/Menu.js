import React, { Component } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import { NavLink } from 'react-router-dom'
import Avatar from './Avatar'

class Menu extends Component{
    state = { showMenu: false }
    toggleMenu(){
        this.setState({ showMenu: !this.state.showMenu })
    }
    generateRoutes(){
        return this.props.routes.map((route) => (
            <NavLink className="link" to={route.href} key={route.href}>
                {route.name}
            </NavLink>
        ))
    }
    renderMenu(){
        return(
            <div className="container">
                {this.generateRoutes()}
                {this.props.isLoggedIn && <Avatar {...this.props.profile}/>}
            </div>
        )
    }
    renderMobileMenu(){
        const { showMenu } = this.state
        const toggleIcon = cx("fa", { "fa-window-close" : showMenu, "fa-bars" : !showMenu })
        return(
            <div id="mobile" className="container">
                <i 
                    onClick={this.toggleMenu.bind(this)}
                    className={toggleIcon}
                />
                <div className="menu">
                    {showMenu && this.generateRoutes()}
                </div>
            </div>
        )
    }
    render(){
        /** Toggles z-index to slide behind content when showMenu is false and overlay menu is active **/
        const smartHideStyle = { zIndex: this.state.showMenu ? 100 : 0 } 
        return(
            <nav style={smartHideStyle} className="navbar">
                {this.renderMenu()}
                {this.renderMobileMenu()}
            </nav>
        )
    }
}

Menu.propTypes = {
    profile: PropTypes.object.isRequired,
    routes: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string.isRequired,
        href: PropTypes.string.isRequired
    })).isRequired,
    isLoggedIn: PropTypes.bool.isRequired
}

export default Menu