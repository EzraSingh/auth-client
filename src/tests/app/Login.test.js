import React from 'react'
import { Redirect } from 'react-router-dom'
import LoginView from 'app/Login'
import { LoginForm } from 'app/Login/components/Form'
import { shallow } from 'enzyme'

describe('login view', () => {
    describe('<Login />', () => {
        it('should render', () => {
            const component = shallow(<LoginView/>)
            expect(component).toMatchSnapshot()
        })
    })
    describe('<LoginForm />', () => {
        const props = { isLoggedIn: false, login: jest.fn() }
        const event = { preventDefault: jest.fn() }
        beforeEach(() => {
            props.isLoggedIn = false
        })
        it('should redirect if user is authenticated', () => {
            props.isLoggedIn = true
            const component = shallow(<LoginForm {...props}/>)
            expect(component.equals(<Redirect to="/profile"/>)).toBe(true)
        })
        it('should submit login request on submit', () => {
            const component = shallow(<LoginForm {...props}/>)
            component.find('form').simulate('submit', event)
            expect(props.login).toHaveBeenCalledTimes(1)
        })
        it('should toggle password form visibility', () => {
            const component = shallow(<LoginForm {...props} />)
            const input = () => component.find('#password')
            const toggle = component.find('i.fas') 
            expect(input().props().type).toBe('password')
            toggle.simulate('click')
            expect(input().props().type).toBe('text')
            toggle.simulate('click')
            expect(input().props().type).toBe('password')
        })
    })
})