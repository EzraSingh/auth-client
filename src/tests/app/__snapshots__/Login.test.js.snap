// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login view <Login /> should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="form box animated fadeInUp"
      >
        <h2>
          Login
        </h2>
        <Connect(LoginForm) />
        <Link
          replace={false}
          to="/register"
        >
          Don't have an account?
        </Link>
      </div>,
      "className": "login overlay flex h-center v-center",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Login
          </h2>,
          <Connect(LoginForm) />,
          <Link
            replace={false}
            to="/register"
          >
            Don't have an account?
          </Link>,
        ],
        "className": "form box animated fadeInUp",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login",
          },
          "ref": null,
          "rendered": "Login",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Don't have an account?",
            "replace": false,
            "to": "/register",
          },
          "ref": null,
          "rendered": "Don't have an account?",
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="form box animated fadeInUp"
        >
          <h2>
            Login
          </h2>
          <Connect(LoginForm) />
          <Link
            replace={false}
            to="/register"
          >
            Don't have an account?
          </Link>
        </div>,
        "className": "login overlay flex h-center v-center",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2>
              Login
            </h2>,
            <Connect(LoginForm) />,
            <Link
              replace={false}
              to="/register"
            >
              Don't have an account?
            </Link>,
          ],
          "className": "form box animated fadeInUp",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Login",
            },
            "ref": null,
            "rendered": "Login",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Don't have an account?",
              "replace": false,
              "to": "/register",
            },
            "ref": null,
            "rendered": "Don't have an account?",
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
